### PR TITLE
feat: add support for npm owner

### PR DESCRIPTION
**Pull request by mbtools**
Here we go... 

### Overview (based on [`npm owner`](https://docs.npmjs.com/cli/v10/commands/npm-owner))

This PR will allow executing the `npm owner` command on a verdaccio registry. 

The current owners of a package are stored using the `maintainers` property of `package.json`. Each version will also contain the historical list of owners at the time of publishing in the `maintainers` property of the version.

*Note 1*

Verdaccio does not store user emails. Therefore, maintainers are added using just the username (and an empty email field). There's also *no email invitation* flow for added owners as with the npm registry. Adding emails to user profiles might be a future enhancement (`npm profile`).

*Note 2*

By default, this PR will *not* impact on permissions for accessing packages. A new config option can be used to activate, checking package ownership when trying publish a new or unpublish an old version of a package:

```
publish:
  check_owners: true
```

If `check_owners` is `true`, then changing a package (publish/unpublish), will require that your user is listed as an owner of the package i.e. included in the `maintainers` array of the packument.

### Technical details (based on [`npm cli`](https://github.com/npm/cli/blob/latest/lib/commands/owner.js))

- `npm owner list`
  - uses `get /:package` endpoint to retrieve the manifest and maintainers

- `npm owner add/rm`
  - uses `get /-/user/:org_couchdb_user` endpoint to get the users name and email (see below)
  - the maintainer list in the manifest is updated using `put /:package` endpoint (via request body)
  - the implementation is very similar to `npm star`

- `get /-/user/:org_couchdb_user`
  - added `{ name: <username>, email: '' }` to the response if user is logged in (required by `npm owner`)
  - changed response to `{ ok: false }` if user is not logged in (to mirror npm registry)

- `put /-/user/:org_couchdb_user/:_rev?/:revision?`
  - added validation of URL and request body to harden endpoint
    - example: `put /-/user/org.couchdb.user:tom` will raise an error when `name: 'jerry'` is contained in the body

- `put /:package`
  - `npm publish` will set `maintainers: [ { name: <logged_in_user>, email: '' }]` for new packages
  - `npm owner` will updated `maintainers` as provided by `npm cli` adding or removing owners

### Status

- [x] `npm owner list`
- [x] `npm owner add/rm`
- [x] `npm publish`, update package
- [x] `npm publish`, update version metadata
- [x] Add tests for the new scenarios

### Example

```bash
> npm config get registry
http://localhost:8000/
> npm owner ls
mbtools <no-reply@marcbernardtools.com>
> npm owner add verdaccio
+ verdaccio (mbtools-demo)
> npm owner ls
mbtools <no-reply@marcbernardtools.com>
verdaccio <>
> npm owner rm verdaccio 
- verdaccio (mbtools-demo)
> npm owner ls
mbtools <no-reply@marcbernardtools.com>
> npm owner rm mbtools  
npm ERR! code EOWNERRM
npm ERR! Cannot remove all owners of a package. Add someone else first.
npm ERR! A complete log of this run can be found in: ...debug-0.log
```

### CODE REVIEW

**Comment by mbtools** [comment]
`npm owner` is now updating the `package.json` in verdaccio the same way as in the npm registry 😀 

Next would be addressing _Note 1_ and _Note 2_ from above.

**Comment by juanpicado** [comment]
> `npm owner` is now updating the `package.json` in verdaccio the same way as in the npm registry 😀
> 
> Next would be addressing _Note 1_ and _Note 2_ from above.

I didn't have much focus time to finish but I have no objections so far, great job I'll back to this next week (I'm often afk lately)

**Comment by mbtools** [comment]
Todos:

- [ ] add `checkAllowedToChangePackage` to storage `removePackage`
- [ ] add `checkAllowedToChangePackage` to storage `removeTarball`

**Comment by mbtools** [comment]
This is it 😃 If you set `publish > check_owners: true`, users will have to be authenticated and belong to the maintainers of a package just like the `npm registry` (without `org` and `team`).

PS: I will update docs in separate PR.

**Comment by juanpicado** [comment]
> This is it 😃 If you set `publish > check_owners: true`, users will have to be authenticated and belong to the maintainers of a package just like the `npm registry` (without `org` and `team`).
> 
> 
> 
> PS: I will update docs in separate PR. 

Still half way to finish the deep testing but yeah that was one of my questions I had so far :) sorry I need to focus on the cipher issue but I will be back on this.

**Comment by mbtools** [comment]
fixing getTarballDetails, then back to this

**Comment by mbtools** [comment]
Rebased. Secrets are related to tests. Just ignore? Will these findings show on all PRs?

**Comment by juanpicado** [comment]
> Rebased. Secrets are related to tests. Just ignore? Will these findings show on all PRs?

solved :) the GitGuardian allows set some files as tests, so I did.

File: packages/store/src/storage.ts [hunk 1]
```diff
@@ -1390,6 +1467,10 @@ class Storage {
       },
     };
 
+    // Set initial package owner
+    // TODO: Add email of user
+    packageData.maintainers = [{ name: username || '', email: '' }];
```
Comment (juanpicado) [inline packages/store/src/storage.ts:1472]: What if `username` is undefined?

**Review by juanpicado — COMMENTED** [review]
Great stuff, I tested thoughtfully and I have only two small request
- Update config files
- Pending question regarding possible undefined. 
- Optional (https://github.com/verdaccio/verdaccio/blob/master/website/docs/config.md#advanced-settings-advanced-settings) only for next version (no 5.x or 6.x yet) I still need to figure out how to replace storage dependency.

**Comment by mbtools** [comment]
Thanks for the feedback! I have busy week speaking at a [conference](https://abapconf.org). I will get back to it afterwards.

File: packages/store/src/storage.ts (hunk 1 above)
Comment (mbtools) [reply packages/store/src/storage.ts:1472]: I changed it to DEFAULT_USER (Anonymous) if undefined

**Comment by mbtools** [comment]
Fixed the undefined user issue and updated the docs/readme 😃

**Pull request merged**